File: java/app/src/main/java/com/fatchickensoftware/books/BooksAPI.java
package com.fatchickensoftware.books;

import com.google.gson.Gson;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class BooksAPI {

    public interface SearchCallback {
        void onResult(Volumes volumes);
    }

    private OkHttpClient client;

    public BooksAPI(OkHttpClient client) {
        this.client = client;
    }

    public void search(String apiKey, String query, int startIndex, int maxResults, SearchCallback callback) {
        HttpUrl.Builder urlBuilder = HttpUrl.parse("https://www.googleapis.com/books/v1/volumes").newBuilder();
        urlBuilder.addQueryParameter("q", query);
        urlBuilder.addQueryParameter("startIndex", Integer.toString(startIndex));
        urlBuilder.addQueryParameter("maxResults", Integer.toString(maxResults));

        Request request = new Request.Builder()
                .addHeader("accept", "application/json")
                .addHeader("key", apiKey)
                .url(urlBuilder.build())
                .build();

        asyncRequest(request, callback);
    }

    private void asyncRequest(Request request, SearchCallback callback) {
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                callback.onResult(null);
            }

            @Override
            public void onResponse(Call call, Response response) {
                if (response.isSuccessful()) {
                    try {
                        callback.onResult(new Gson().fromJson(response.body().string(), Volumes.class));
                    } catch (IOException e) {
                        e.printStackTrace();
                        callback.onResult(null);
                    }
                } else {
                    callback.onResult(null);
                }
            }
        });
    }
}

File: kotlin/app/src/main/java/com/fatchickensoftware/books/BooksAPI.kt
package com.fatchickensoftware.books

import com.google.gson.Gson
import okhttp3.*
import java.io.IOException
import java.io.Serializable

data class ImageLinks (
    val smallThumbnail: String? = null,
    val thumbnail: String? = null
) : Serializable

data class VolumeInfo(
    val title: String? = null,
    val subtitle: String? = null,
    val description: String? = null,
    val authors: List<String> = emptyList(),
    val publisher: String? = null,
    val publishedDate: String? = null,
    val imageLinks: ImageLinks? = null
) : Serializable

data class Volume(
    val id: String = "",
    val volumeInfo: VolumeInfo? = null
) : Serializable

data class Volumes(
    val items: List<Volume> = emptyList()
)

typealias SearchCallback = (Volumes?) -> Unit

class BooksAPI(private val client: OkHttpClient) {

    fun search(query: String, startIndex: Int, maxResults: Int, callback: SearchCallback) {
        val url = HttpUrl.Builder()
            .scheme("https")
            .host("www.googleapis.com")
            .addPathSegment("books")
            .addPathSegment("v1")
            .addPathSegment("volumes")
            .addQueryParameter("q", query)
            .addQueryParameter("startIndex", startIndex.toString())
            .addQueryParameter("maxResults", maxResults.toString())
            .build()

        val request = Request.Builder()
            .addHeader("accept", "application/json")
            .addHeader("key", BOOKS_API_KEY)
            .url(url)
            .build()

        asyncRequest(request, callback)
    }

    private inline fun <reified T> asyncRequest(request: Request, crossinline callback: (T?) -> Unit) {
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(null)
            }

            override fun onResponse(call: Call, response: Response) {
                when {
                    response.isSuccessful -> callback(Gson().fromJson(response.body()?.string(), T::class.java))
                    else -> callback(null)
                }
            }
        })
    }
}
